<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>batch_mask.pl</title>
<link rel="stylesheet" href="../css/dawgpaws_man.css" type="text/css" />
<link rev="made" href="mailto:root@localhost" />
</head>

<body>
<table border="0" width="100%" cellspacing="0" cellpadding="3">
<tr><td class="block" valign="middle">
<big><strong><span class="block">&nbsp;batch_mask.pl</span></strong></big>
</td></tr>
</table>

<p><a name="__index__"></a></p>
<!-- INDEX BEGIN -->

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#version">VERSION</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<ul>

		<li><a href="#usage">Usage</a></li>
		<li><a href="#required_arguments">Required Arguments</a></li>
	</ul>

	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#required_arguments">Required Arguments</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#diagnostics">DIAGNOSTICS</a></li>
	<li><a href="#configuration_and_environment">CONFIGURATION AND ENVIRONMENT</a></li>
	<ul>

		<li><a href="#configuration_file">Configuration File</a></li>
	</ul>

	<li><a href="#dependencies">DEPENDENCIES</a></li>
	<ul>

		<li><a href="#required_software">Required Software</a></li>
		<li><a href="#required_perl_modules">Required Perl Modules</a></li>
	</ul>

	<li><a href="#bugs_and_limitations">BUGS AND LIMITATIONS</a></li>
	<ul>

		<li><a href="#bugs">Bugs</a></li>
		<li><a href="#limitations">Limitations</a></li>
	</ul>

	<li><a href="#see_also">SEE ALSO</a></li>
	<li><a href="#license">LICENSE</a></li>
	<li><a href="#author">AUTHOR</a></li>
	<li><a href="#history">HISTORY</a></li>
</ul>
<!-- INDEX END -->

<hr />
<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>batch_mask.pl - Run RepeatMasker and parse results to a GFF format file.</p>
<p>
</p>
<hr />
<h1><a name="version">VERSION</a></h1>
<p>This documentation refers to program version $Rev: 323 $</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>
</p>
<h2><a name="usage">Usage</a></h2>
<pre>
    batch_mask.pl -i DirToProcess -o OutDir -c ConfigFile</pre>
<p>
</p>
<h2><a name="required_arguments">Required Arguments</a></h2>
<pre>
    -i, --indir    # Directory of fasta files to process
    -o, --outdir   # Path to the base output directory
    -c, --config   # Path to the config file</pre>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>Runs the RepeatMasker program for a set of input
FASTA files against a set of repeat library files &amp;
then converts the repeat masker *.out file into the
GFF format and then to the game XML format for
visualization by the Apollo genome anotation program.</p>
<p>
</p>
<hr />
<h1><a name="required_arguments">Required Arguments</a></h1>
<dl>
<dt><strong><a name="item_%2di%2c%2d%2dindir">-i,--indir</a></strong><br />
</dt>
<dd>
Path of the directory containing the sequences to process.
</dd>
<p></p>
<dt><strong><a name="item_%2do%2c%2d%2doutdir">-o,--outdir</a></strong><br />
</dt>
<dd>
Path of the directory to place the program output.
</dd>
<p></p>
<dt><strong><a name="item_%2dc%2c_%2d%2dconfig">-c, --config</a></strong><br />
</dt>
<dd>
Configuration file that lists the database names and paths of the
fasta files to use as masking databases.
</dd>
<p></p></dl>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<dl>
<dt><strong><a name="item_%2dp%2c%2d%2dnum%2dproc">-p,--num-proc</a></strong><br />
</dt>
<dd>
The number of processors to use for RepeatMasker. Default is one.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dengine">--engine</a></strong><br />
</dt>
<dd>
The repeatmasker engine to use: [crossmatch|wublast|decypher].
The default is to use the crossmatch engine.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dapollo">--apollo</a></strong><br />
</dt>
<dd>
Use the apollo program to convert the file from gff to game xml.
The default is not to use apollo.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2drm%2dpath">--rm-path</a></strong><br />
</dt>
<dd>
The full path to the RepeatMasker binary.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dlogfile">--logfile</a></strong><br />
</dt>
<dd>
Path to a file that will be used to log program status.
If the file already exists, additional information will be concatenated
to the existing file.
</dd>
<p></p>
<dt><strong><a name="item_%2dq%2c%2d%2dquiet">-q,--quiet</a></strong><br />
</dt>
<dd>
Run the program with minimal output.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dtest">--test</a></strong><br />
</dt>
<dd>
Run the program without doing the system commands.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dusage">--usage</a></strong><br />
</dt>
<dd>
Short overview of how to use program from command line.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dhelp">--help</a></strong><br />
</dt>
<dd>
Show program usage with summary of options.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dversion">--version</a></strong><br />
</dt>
<dd>
Show program version.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dman">--man</a></strong><br />
</dt>
<dd>
Show the full program manual. This uses the perldoc command to print the 
POD documentation for the program.
</dd>
<p></p></dl>
<p>
</p>
<hr />
<h1><a name="diagnostics">DIAGNOSTICS</a></h1>
<p>Error messages generated by this program and possible solutions are listed
below.</p>
<dl>
<dt><strong><a name="item_error%3a_no_fasta_files_were_found_in_the_input_di">ERROR: No fasta files were found in the input directory</a></strong><br />
</dt>
<dd>
The input directory does not contain fasta files in the expected format.
This could happen because you gave an incorrect path or because your sequence 
files do not have the expected *.fasta extension in the file name.
</dd>
<p></p>
<dt><strong><a name="item_error%3a_could_not_create_the_output_directory">ERROR: Could not create the output directory</a></strong><br />
</dt>
<dd>
The output directory could not be created at the path you specified. 
This could be do to the fact that the directory that you are trying
to place your base directory in does not exist, or because you do not
have write permission to the directory you want to place your file in.
</dd>
<p></p></dl>
<p>
</p>
<hr />
<h1><a name="configuration_and_environment">CONFIGURATION AND ENVIRONMENT</a></h1>
<p>
</p>
<h2><a name="configuration_file">Configuration File</a></h2>
<p>The major configuration file for this program is the list of
datbases indicated by the -c flag.</p>
<p>This file is a tab delimited text file. 
Lines beginning with # are ignored.</p>
<p><strong>EXAMPLE</strong></p>
<pre>
  #-------------------------------------------------------------
  # DBNAME       DB_PATH
  #-------------------------------------------------------------
  TREP_9         /db/repeats/Trep9.nr.fasta
  TIGR_Trit      /db/repeats/TIGR_Triticum_GSS_Repeats.v2.fasta
  # END</pre>
<p>The columns above represent the following</p>
<dl>
<dt><strong><a name="item_col%2e_1">Col. 1</a></strong><br />
</dt>
<dd>
The name of the repeat library
This will be used to name the output files from the analysis
and to name the data tracks that will be used by Apollo.
</dd>
<p></p>
<dt><strong><a name="item_col%2e_2">Col. 2</a></strong><br />
</dt>
<dd>
The path to the fasta format file containing the repeats.
</dd>
<p></p></dl>
<p>
</p>
<hr />
<h1><a name="dependencies">DEPENDENCIES</a></h1>
<p>
</p>
<h2><a name="required_software">Required Software</a></h2>
<ul>
<li><strong><a name="item_repeatmasker">RepeatMasker</a></strong><br />
</li>
The RepeatMasker program can be download at:
(http://www.repeatmasker.org/)
<p></p>
<li><strong><a name="item_apollo_genome_annotation_curation_tool">Apollo Genome Annotation Curation Tool</a></strong><br />
</li>
The Apollo Genome Annotation Curation tool can be downloaded at
<a href="http://www.fruitfly.org/annot/apollo/">http://www.fruitfly.org/annot/apollo/</a> .
<p></p></ul>
<p>
</p>
<h2><a name="required_perl_modules">Required Perl Modules</a></h2>
<ul>
<li><strong><a name="item_file%3a%3acopy">File::Copy</a></strong><br />
</li>
This module is required to copy the BLAST results.
<p></p>
<li><strong><a name="item_getopt%3a%3along">Getopt::Long</a></strong><br />
</li>
This module is required to accept options at the command line.
<p></p></ul>
<p>
</p>
<hr />
<h1><a name="bugs_and_limitations">BUGS AND LIMITATIONS</a></h1>
<p>
</p>
<h2><a name="bugs">Bugs</a></h2>
<ul>
<li><strong><a name="item_no_bugs_currently_known">No bugs currently known</a></strong><br />
</li>
If you find a bug with this software, file a bug report on the DAWG-PAWS
Sourceforge website: <a href="http://sourceforge.net/tracker/?group_id=204962">http://sourceforge.net/tracker/?group_id=204962</a>
<p></p></ul>
<p>
</p>
<h2><a name="limitations">Limitations</a></h2>
<ul>
<li><strong><a name="item_limited_repeatmasker_version_testing">Limited RepeatMasker version testing</a></strong><br />
</li>
This program has been tested with RepeatMasker v  3.1.6
<p></p>
<li><strong><a name="item_no_env_options">No Env Options</a></strong><br />
</li>
Currently this program does not make use of variables in the user
environment. However, it would be useful to define program paths
and some common options in the environment.
<p></p></ul>
<p>
</p>
<hr />
<h1><a name="see_also">SEE ALSO</a></h1>
<p>The batch_mask.pl program is part of the DAWG-PAWS package of genome
annotation programs. See the DAWG-PAWS web page 
( <a href="http://dawgpaws.sourceforge.net/">http://dawgpaws.sourceforge.net/</a> )
or the Sourceforge project page 
( <a href="http://sourceforge.net/projects/dawgpaws">http://sourceforge.net/projects/dawgpaws</a> ) 
for additional information about this package.</p>
<p>
</p>
<hr />
<h1><a name="license">LICENSE</a></h1>
<p>GNU GENERAL PUBLIC LICENSE, VERSION 3</p>
<p><a href="http://www.gnu.org/licenses/gpl.html">http://www.gnu.org/licenses/gpl.html</a></p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<p>James C. Estill &lt;JamesEstill at gmail.com&gt;</p>
<p>
</p>
<hr />
<h1><a name="history">HISTORY</a></h1>
<p>STARTED: 04/10/2006

</p>
<p>UPDATED: 12/10/2007

</p>
<p>VERSION: $Rev: 323 $

</p>
<table border="0" width="100%" cellspacing="0" cellpadding="3">
<tr><td class="block" valign="middle">
<big><strong><span class="block">&nbsp;batch_mask.pl</span></strong></big>
</td></tr>
</table>

</body>

</html>
