<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>batch_findgaps.html</title>
<link rel="stylesheet" href="../css/dawgpaws_man.css" type="text/css" />
<link rev="made" href="mailto:root@b05.apple.com" />
</head>

<body>
<table border="0" width="100%" cellspacing="0" cellpadding="3">
<tr><td class="block" valign="middle">
<big><strong><span class="block">&nbsp;batch_findgaps.html</span></strong></big>
</td></tr>
</table>

<p><a name="__index__"></a></p>
<!-- INDEX BEGIN -->

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#version">VERSION</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<ul>

		<li><a href="#usage">Usage</a></li>
		<li><a href="#required_arguments">Required Arguments</a></li>
	</ul>

	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#required_arguments">REQUIRED ARGUMENTS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#diagnostics">DIAGNOSTICS</a></li>
	<li><a href="#configuration_and_environment">CONFIGURATION AND ENVIRONMENT</a></li>
	<li><a href="#dependencies">DEPENDENCIES</a></li>
	<ul>

		<li><a href="#required_software">Required Software</a></li>
		<li><a href="#required_perl_modules">Required Perl Modules</a></li>
	</ul>

	<li><a href="#bugs_and_limitations">BUGS AND LIMITATIONS</a></li>
	<ul>

		<li><a href="#bugs">Bugs</a></li>
		<li><a href="#limitations">Limitations</a></li>
	</ul>

	<li><a href="#see_also">SEE ALSO</a></li>
	<li><a href="#license">LICENSE</a></li>
	<li><a href="#author">AUTHOR</a></li>
	<li><a href="#history">HISTORY</a></li>
</ul>
<!-- INDEX END -->

<hr />
<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>batch_findgaps.pl - Annotate gaps in a fasta file</p>
<p>
</p>
<hr />
<h1><a name="version">VERSION</a></h1>
<p>This documentation refers to batch_findgaps version $Rev: 308 $</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>
</p>
<h2><a name="usage">Usage</a></h2>
<pre>
    batch_findgaps.pl -i DirToProcess -o OutDir</pre>
<p>
</p>
<h2><a name="required_arguments">Required Arguments</a></h2>
<pre>
    -i, --indir    # Directory of fasta files to process
    -o, --outdir   # Path to the base output directory</pre>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>Runs the RepeatMasker program for a set of input
FASTA files against a set of repeat library files &amp;
then converts the repeat masker *.out file into the
GFF format and then to the game XML format for
visualization by the Apollo genome anotation program.</p>
<p>
</p>
<hr />
<h1><a name="required_arguments">REQUIRED ARGUMENTS</a></h1>
<dl>
<dt><strong><a name="item__2di_2c_2d_2dindir">-i,--indir</a></strong><br />
</dt>
<dd>
Path of the directory containing the sequences to process.
</dd>
<p></p>
<dt><strong><a name="item__2do_2c_2d_2doutdir">-o,--outdir</a></strong><br />
</dt>
<dd>
Path of the directory to place the program output.
</dd>
<p></p></dl>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<dl>
<dt><strong><a name="item__2d_2dm_2cmask">--m,mask</a></strong><br />
</dt>
<dd>
Single letter to mask with. Valid options are: [ N | n | X | x ]
</dd>
<p></p>
<dt><strong><a name="item__2d_2dext">--ext</a></strong><br />
</dt>
<dd>
The new outfile extension to use. Default value is .hard.fasta
</dd>
<p></p>
<dt><strong><a name="item__2d_2dlogfile">--logfile</a></strong><br />
</dt>
<dd>
Path to a file that will be used to log program status.
If the file already exists, additional information will be concatenated
to the existing file.
</dd>
<p></p>
<dt><strong><a name="item__2d_2dusage">--usage</a></strong><br />
</dt>
<dd>
Short overview of how to use program from command line.
</dd>
<p></p>
<dt><strong><a name="item__2d_2dhelp">--help</a></strong><br />
</dt>
<dd>
Show program usage with summary of options.
</dd>
<p></p>
<dt><strong><a name="item__2d_2dversion">--version</a></strong><br />
</dt>
<dd>
Show program version.
</dd>
<p></p>
<dt><strong><a name="item__2d_2dman">--man</a></strong><br />
</dt>
<dd>
Show the full program manual. This uses the perldoc command to print the 
POD documentation for the program.
</dd>
<p></p>
<dt><strong><a name="item__2dq_2c_2d_2dquiet">-q,--quiet</a></strong><br />
</dt>
<dd>
Run the program with minimal output.
</dd>
<p></p>
<dt><strong><a name="item__2d_2dtest">--test</a></strong><br />
</dt>
<dd>
Run the program without doing the system commands.
</dd>
<p></p></dl>
<p>
</p>
<hr />
<h1><a name="diagnostics">DIAGNOSTICS</a></h1>
<dl>
<dt><strong><a name="item_error_3a_no_fasta_files_were_found_in_the_input_di">ERROR: No fasta files were found in the input directory</a></strong><br />
</dt>
<dd>
The input directory does not contain fasta files in the expected format.
This could happen because you gave an incorrect path or because your sequence 
files do not have the expected *.fasta extension in the file name.
</dd>
<p></p>
<dt><strong><a name="item_error_3a_could_not_create_the_output_directory">ERROR: Could not create the output directory</a></strong><br />
</dt>
<dd>
The output directory could not be created at the path you specified. 
This could be do to the fact that the directory that you are trying
to place your base directory in does not exist, or because you do not
have write permission to the directory you want to place your file in.
</dd>
<p></p></dl>
<p>
</p>
<hr />
<h1><a name="configuration_and_environment">CONFIGURATION AND ENVIRONMENT</a></h1>
<p>No configuration files or environmental variables are required to use
this program.</p>
<p>
</p>
<hr />
<h1><a name="dependencies">DEPENDENCIES</a></h1>
<p>
</p>
<h2><a name="required_software">Required Software</a></h2>
<p>Additional software is not required to use this program.</p>
<p>
</p>
<h2><a name="required_perl_modules">Required Perl Modules</a></h2>
<ul>
<li><strong><a name="item_file_3a_3acopy">File::Copy</a></strong><br />
</li>
This module is required to copy the BLAST results.
<p></p>
<li><strong><a name="item_getopt_3a_3along">Getopt::Long</a></strong><br />
</li>
This module is required to accept options at the command line.
<p></p></ul>
<p>
</p>
<hr />
<h1><a name="bugs_and_limitations">BUGS AND LIMITATIONS</a></h1>
<p>
</p>
<h2><a name="bugs">Bugs</a></h2>
<ul>
<li><strong><a name="item_incorrect_end_position_of_annotated_gap">Incorrect end position of annotated gap</a></strong><br />
</li>
Early versions of this script would assign an incorrect end position the
the 3' end of a gap.
<p></p>
<li><strong><a name="item_bug_reporting">Bug Reporting</a></strong><br />
</li>
If you find a bug with this software, file a bug report on the DAWG-PAWS
Sourceforge website: <a href="http://sourceforge.net/tracker/?group_id=204962">http://sourceforge.net/tracker/?group_id=204962</a>
<p></p></ul>
<p>
</p>
<h2><a name="limitations">Limitations</a></h2>
<ul>
<li><strong><a name="item_recognized_gap_characters">Recognized gap characters</a></strong><br />
</li>
Due to the way that regular expressions are coded in PERL, the characters
that can be used to indicate gaps must be hard coded. The characters that are
currently hard coded for recognition by batch_findgaps are n,N,x, and X. If 
there are additional characters you would like to add as a recognized
gap character, file a Feature Request on the DAWG-PAWS poject page on
Sourceforge ( <a href="http://sourceforge.net/tracker/?group_id=204962&atid=991722">http://sourceforge.net/tracker/?group_id=204962&atid=991722</a> ).
<p></p>
<li><strong><a name="item_limited_file_extensions_are_supported">Limited file extensions are supported</a></strong><br />
</li>
BLAST output file must currently end with blo, bln, or blx. For example
a BLASTx output may be named BlastOut.blx while a BLASTN output
may be names BlastOut.bln. FASTA files must end with a fasta or fa extension.
For examples must have names like my_seq.fasta or my_seq.fa.
<p></p></ul>
<p>
</p>
<hr />
<h1><a name="see_also">SEE ALSO</a></h1>
<p>The batch_blast.pl program is part of the DAWG-PAWS package of genome
annotation programs. See the DAWG-PAWS web page 
( <a href="http://dawgpaws.sourceforge.net/">http://dawgpaws.sourceforge.net/</a> )
or the Sourceforge project page 
( <a href="http://sourceforge.net/projects/dawgpaws">http://sourceforge.net/projects/dawgpaws</a> ) 
for additional information about this package.</p>
<p>
</p>
<hr />
<h1><a name="license">LICENSE</a></h1>
<p>GNU GENERAL PUBLIC LICENSE, VERSION 3</p>
<p><a href="http://www.gnu.org/licenses/gpl.html">http://www.gnu.org/licenses/gpl.html</a></p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<p>James C. Estill &lt;JamesEstill at gmail.com&gt;</p>
<p>
</p>
<hr />
<h1><a name="history">HISTORY</a></h1>
<p>STARTED: 08/01/2007</p>
<p>UPDATED: 12/06/2007</p>
<p>VERSION: $Rev: 308 $</p>
<table border="0" width="100%" cellspacing="0" cellpadding="3">
<tr><td class="block" valign="middle">
<big><strong><span class="block">&nbsp;batch_findgaps.html</span></strong></big>
</td></tr>
</table>

</body>

</html>
