<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>batch_genemark.pl</title>
<link rel="stylesheet" href="../css/dawgpaws_man.css" type="text/css" />
<link rev="made" href="mailto:root@localhost" />
</head>

<body>
<table border="0" width="100%" cellspacing="0" cellpadding="3">
<tr><td class="block" valign="middle">
<big><strong><span class="block">&nbsp;batch_genemark.pl</span></strong></big>
</td></tr>
</table>

<p><a name="__index__"></a></p>
<!-- INDEX BEGIN -->

<ul>

	<li><a href="#version">VERSION</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<ul>

		<li><a href="#usage">Usage</a></li>
		<li><a href="#required_arguments">Required Arguments</a></li>
	</ul>

	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#required_arguments">REQUIRED ARGUMENTS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#diagnostics">DIAGNOSTICS</a></li>
	<li><a href="#configuration_and_environment">CONFIGURATION AND ENVIRONMENT</a></li>
	<ul>

		<li><a href="#configuration_file">Configuration File</a></li>
		<li><a href="#user_environment">User Environment</a></li>
	</ul>

	<li><a href="#dependencies">DEPENDENCIES</a></li>
	<ul>

		<li><a href="#required_software">Required Software</a></li>
		<li><a href="#required_perl_modules">Required Perl Modules</a></li>
	</ul>

	<li><a href="#bugs_and_limitations">BUGS AND LIMITATIONS</a></li>
	<ul>

		<li><a href="#bugs">Bugs</a></li>
		<li><a href="#limitations">Limitations</a></li>
	</ul>

	<li><a href="#see_also">SEE ALSO</a></li>
	<li><a href="#license">LICENSE</a></li>
	<li><a href="#author">AUTHOR</a></li>
	<li><a href="#history">HISTORY</a></li>
</ul>
<!-- INDEX END -->

<hr />
<p>
</p>
<h1><a name="version">VERSION</a></h1>
<p>This documentation refers to batch_genemark.pl version $Rev: 368 $</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>
</p>
<h2><a name="usage">Usage</a></h2>
<pre>
    batch_genscan.pl -i DirToProcess -o OutDir</pre>
<p>
</p>
<h2><a name="required_arguments">Required Arguments</a></h2>
<pre>
    -i, --indir    # Directory of fasta files to process
    -o, --outdir   # Path to the base output directory
    -c, --config   # Path to the config file</pre>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>Run the GeneMarkHMM gene prediction program in batch mode.
Runs genmark as well as converts output to gff format.
Requires a config file to specify libraries to use.</p>
<p>
</p>
<hr />
<h1><a name="required_arguments">REQUIRED ARGUMENTS</a></h1>
<dl>
<dt><strong><a name="item_%2di%2c%2d%2dindir">-i,--indir</a></strong><br />
</dt>
<dd>
Path of the directory containing the sequences to process.
</dd>
<p></p>
<dt><strong><a name="item_%2do%2c%2d%2doutdir">-o,--outdir</a></strong><br />
</dt>
<dd>
Path of the directory to place the program output.
</dd>
<p></p>
<dt><strong><a name="item_%2dc%2c_%2d%2dconfig">-c, --config</a></strong><br />
</dt>
<dd>
Currently this program does NOT make use of a configuration file. This will
be a fairly easy thing to add, and is on my TODO List.
</dd>
<p></p></dl>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<dl>
<dt><strong><a name="item_%2d%2dgenemark%2ddir">--genemark-dir</a></strong><br />
</dt>
<dd>
Directory that contains the GeneMark.hmm binaries. This can also be set
with the environment variable GM_BIN_DIR.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dlib%2ddir">--lib-dir</a></strong><br />
</dt>
<dd>
The full path to the directory that contains the model libraries
for GeneMarkHMM. This can also be set with the environment varaible
GM_LIB_DIR.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dusage">--usage</a></strong><br />
</dt>
<dd>
Short overview of how to use program from command line.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dhelp">--help</a></strong><br />
</dt>
<dd>
Show program usage with summary of options.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dversion">--version</a></strong><br />
</dt>
<dd>
Show program version.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dman">--man</a></strong><br />
</dt>
<dd>
Show the full program manual. This uses the perldoc command to print the 
POD documentation for the program.
</dd>
<p></p>
<dt><strong><a name="item_%2dq%2c%2d%2dquiet">-q,--quiet</a></strong><br />
</dt>
<dd>
Run the program with minimal output.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dtest">--test</a></strong><br />
</dt>
<dd>
Run the program without doing the system commands.
</dd>
<p></p></dl>
<p>
</p>
<hr />
<h1><a name="diagnostics">DIAGNOSTICS</a></h1>
<p>Error messages generated by this program and possible solutions are listed
below.</p>
<dl>
<dt><strong><a name="item_error%3a_no_fasta_files_were_found_in_the_input_di">ERROR: No fasta files were found in the input directory</a></strong><br />
</dt>
<dd>
The input directory does not contain fasta files in the expected format.
This could happen because you gave an incorrect path or because your sequence 
files do not have the expected *.fasta extension in the file name.
</dd>
<p></p>
<dt><strong><a name="item_error%3a_could_not_create_the_output_directory">ERROR: Could not create the output directory</a></strong><br />
</dt>
<dd>
The output directory could not be created at the path you specified. 
This could be do to the fact that the directory that you are trying
to place your base directory in does not exist, or because you do not
have write permission to the directory you want to place your file in.
</dd>
<p></p></dl>
<p>
</p>
<hr />
<h1><a name="configuration_and_environment">CONFIGURATION AND ENVIRONMENT</a></h1>
<p>
</p>
<h2><a name="configuration_file">Configuration File</a></h2>
<p>The batch_genemark.pl program does not currently make use of a configuration
file.</p>
<p>
</p>
<h2><a name="user_environment">User Environment</a></h2>
<p>This program makes use of the following variables defined in the
user's environment.</p>
<dl>
<dt><strong><a name="item_gm_bin_dir">GM_BIN_DIR</a></strong><br />
</dt>
<dd>
Directory that contains the GeneMark.hmm binaries.
</dd>
<p></p>
<dt><strong><a name="item_gm_lib_dir">GM_LIB_DIR</a></strong><br />
</dt>
<dd>
The full path to the directory that contains the model libraries
for GeneMarkHMM.
</dd>
<p></p></dl>
<p>The following example illustrates the ENV options set in the bash
shell.</p>
<pre>
    export GM_BIN_DIR='$HOME/apps/GenMark/genemark_hmm_euk.linux/'
    export GM_LIB_DIR='$HOME/apps/GenMark/genemark_hmm_euk.linux/'</pre>
<p>
</p>
<hr />
<h1><a name="dependencies">DEPENDENCIES</a></h1>
<p>
</p>
<h2><a name="required_software">Required Software</a></h2>
<ul>
<li><strong><a name="item_genemark%2ehmm">GeneMark.HMM</a></strong><br />
</li>
The GeneMark.HMM program is available for non-commercial Academic use
for a limited time by applying for a license at:
The <a href="http://opal.biology.gatech/edu/GeneMark">http://opal.biology.gatech/edu/GeneMark</a>
<p></p></ul>
<p>
</p>
<h2><a name="required_perl_modules">Required Perl Modules</a></h2>
<ul>
<li><strong><a name="item_file%3a%3acopy">File::Copy</a></strong><br />
</li>
This module is required to copy the output results.
<p></p>
<li><strong><a name="item_getopt%3a%3along">Getopt::Long</a></strong><br />
</li>
This module is required to accept options at the command line.
<p></p>
<li><strong><a name="item_bio%3a%3atools%3a%3agenemark">Bio::Tools::Genemark</a></strong><br />
</li>
This module is required to parse the results from the Genemark program
The module is part of the BioPerl package <a href="http://www.bioperl.org.">http://www.bioperl.org.</a>
<p></p></ul>
<p>
</p>
<hr />
<h1><a name="bugs_and_limitations">BUGS AND LIMITATIONS</a></h1>
<p>
</p>
<h2><a name="bugs">Bugs</a></h2>
<ul>
<li><strong><a name="item_no_bugs_currently_known">No bugs currently known</a></strong><br />
</li>
If you find a bug with this software, file a bug report on the DAWG-PAWS
Sourceforge website: <a href="http://sourceforge.net/tracker/?group_id=204962">http://sourceforge.net/tracker/?group_id=204962</a>
<p></p></ul>
<p>
</p>
<h2><a name="limitations">Limitations</a></h2>
<ul>
<li><strong><a name="item_limited_gene_model_supported">Limited gene model supported</a></strong><br />
</li>
This program is currently limited to using the gene models that are
relevant to wheat annotation (Rice|Maize|Wheat|Barley). I will be adding
a config file option that will allow multiple gene models to be used.
<p></p></ul>
<p>
</p>
<hr />
<h1><a name="see_also">SEE ALSO</a></h1>
<p>The batch_genemark.pl program is part of the DAWG-PAWS package of genome
annotation programs. See the DAWG-PAWS web page 
( <a href="http://dawgpaws.sourceforge.net/">http://dawgpaws.sourceforge.net/</a> )
or the Sourceforge project page 
( <a href="http://sourceforge.net/projects/dawgpaws">http://sourceforge.net/projects/dawgpaws</a> ) 
for additional information about this package.</p>
<p>
</p>
<hr />
<h1><a name="license">LICENSE</a></h1>
<p>GNU GENERAL PUBLIC LICENSE, VERSION 3</p>
<p><a href="http://www.gnu.org/licenses/gpl.html">http://www.gnu.org/licenses/gpl.html</a></p>
<p>THIS SOFTWARE COMES AS IS, WITHOUT ANY EXPRESS OR IMPLIED
WARRANTY. USE AT YOUR OWN RISK.</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<p>James C. Estill &lt;JamesEstill at gmail.com&gt;</p>
<p>
</p>
<hr />
<h1><a name="history">HISTORY</a></h1>
<p>STARTED: 11/09/2007</p>
<p>UPDATED: 12/14/2007

</p>
<p>VERSION: $Rev: 368 $

</p>
<table border="0" width="100%" cellspacing="0" cellpadding="3">
<tr><td class="block" valign="middle">
<big><strong><span class="block">&nbsp;batch_genemark.pl</span></strong></big>
</td></tr>
</table>

</body>

</html>
