<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>cnv_ta2ap.pl</title>
<link rel="stylesheet" href="../css/dawgpaws_man.css" type="text/css" />
<link rev="made" href="mailto:root@localhost" />
</head>

<body>
<table border="0" width="100%" cellspacing="0" cellpadding="3">
<tr><td class="block" valign="middle">
<big><strong><span class="block">&nbsp;cnv_ta2ap.pl</span></strong></big>
</td></tr>
</table>

<p><a name="__index__"></a></p>
<!-- INDEX BEGIN -->

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<ul>

		<li><a href="#usage">Usage</a></li>
		<li><a href="#required_arguments">Required Arguments</a></li>
	</ul>

	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#required_arguments">REQUIRED ARGUMENTS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#diagnostics">DIAGNOSTICS</a></li>
	<li><a href="#configuration_and_environment">CONFIGURATION AND ENVIRONMENT</a></li>
	<li><a href="#dependencies">DEPENDENCIES</a></li>
	<ul>

		<li><a href="#required_software">Required Software</a></li>
		<li><a href="#required_perl_modules">Required Perl Modules</a></li>
	</ul>

	<li><a href="#bugs_and_limitations">BUGS AND LIMITATIONS</a></li>
	<ul>

		<li><a href="#bugs">Bugs</a></li>
		<li><a href="#limitations">Limitations</a></li>
	</ul>

	<li><a href="#see_also">SEE ALSO</a></li>
	<li><a href="#license">LICENSE</a></li>
	<li><a href="#author">AUTHOR</a></li>
	<li><a href="#history">HISTORY</a></li>
</ul>
<!-- INDEX END -->

<hr />
<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>cnv_ta2ap.pl - Convert TriAnnot GFF3 files to Apollo format</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>
</p>
<h2><a name="usage">Usage</a></h2>
<pre>
    cnv_ta2ap.pl -i InFastaFile -g GffDir -o OutputDir</pre>
<p>
</p>
<h2><a name="required_arguments">Required Arguments</a></h2>
<pre>
    -g, --gffdir   # Directory containing TriAnnot gff files
    -o, --outdir   # Path to the base output directory</pre>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>Converts gff data tracks from gff format to the game
xml format for use in the Apollo Genome Annotation
Curation program. This is for use with output from the
TriAnnot pipeline.</p>
<p>
</p>
<hr />
<h1><a name="required_arguments">REQUIRED ARGUMENTS</a></h1>
<dl>
<dt><strong><a name="item_%2dg%2c%2d%2dgffdir">-g,--gffdir</a></strong><br />
</dt>
<dd>
This is the directory that contains the gff output from the TriAnnot
program that you want to convert.
</dd>
<p></p>
<dt><strong><a name="item_%2do%2c%2d%2doutdir">-o,--outdir</a></strong><br />
</dt>
<dd>
Path of the directory to place the program output.
</dd>
<p></p></dl>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<dl>
<dt><strong><a name="item_%2d%2dcatout">--catout</a></strong><br />
</dt>
<dd>
Optional path to a single gff output file that will contain all
of the converted files concatenated into a single gff file.
</dd>
<p></p>
<dt><strong><a name="item_%2di%2c%2d%2dinfile">-i,--infile</a></strong><br />
</dt>
<dd>
Input FASTA file. The input fasta file will be used as the base file
to map gff sequence features onto when converting from gff to
game.xml output.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dap%2dpath">--ap-path</a></strong><br />
</dt>
<dd>
Full path to the binary file for the Apollo Genome Annotation
Curation program.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dusage">--usage</a></strong><br />
</dt>
<dd>
Short overview of how to use program from command line.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dhelp">--help</a></strong><br />
</dt>
<dd>
Show program usage with summary of options.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dversion">--version</a></strong><br />
</dt>
<dd>
Show program version.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dman">--man</a></strong><br />
</dt>
<dd>
Show the full program manual. This uses the perldoc command to print the 
POD documentation for the program.
</dd>
<p></p>
<dt><strong><a name="item_%2d%2dverbose">--verbose</a></strong><br />
</dt>
<dd>
Run the program with maximum output.
</dd>
<p></p>
<dt><strong><a name="item_%2dq%2c%2d%2dquiet">-q,--quiet</a></strong><br />
</dt>
<dd>
Run the program with minimal output.
</dd>
<p></p></dl>
<p>
</p>
<hr />
<h1><a name="diagnostics">DIAGNOSTICS</a></h1>
<p>Error messages generated by this program and possible solutions are listed
below.</p>
<dl>
<dt><strong><a name="item_error%3a_could_not_create_the_output_directory">ERROR: Could not create the output directory</a></strong><br />
</dt>
<dd>
The output directory could not be created at the path you specified. 
This could be do to the fact that the directory that you are trying
to place your base directory in does not exist, or because you do not
have write permission to the directory you want to place your file in.
</dd>
<p></p></dl>
<p>
</p>
<hr />
<h1><a name="configuration_and_environment">CONFIGURATION AND ENVIRONMENT</a></h1>
<p>The program cnv_ta2ap.pl does not currently use an external configuration
file or make use of variables set the user's environment.</p>
<p>
</p>
<hr />
<h1><a name="dependencies">DEPENDENCIES</a></h1>
<p>
</p>
<h2><a name="required_software">Required Software</a></h2>
<ul>
<li><strong><a name="item_apollo_genome_annotation_tool">Apollo Genome Annotation Tool</a></strong><br />
</li>
To convert output from gff output to the game.xml file format you
will need to have a version of Apollo Genome Annotation program
on your local machine AND have access to the Apollo GUI. Apollo 
can be downloaded at:
<a href="http://www.fruitfly.org/annot/apollo/">http://www.fruitfly.org/annot/apollo/</a>
<p></p></ul>
<p>
</p>
<h2><a name="required_perl_modules">Required Perl Modules</a></h2>
<ul>
<li><strong><a name="item_getopt%3a%3along">Getopt::Long</a></strong><br />
</li>
This module is required to accept options at the command line.
<p></p></ul>
<p>
</p>
<hr />
<h1><a name="bugs_and_limitations">BUGS AND LIMITATIONS</a></h1>
<p>
</p>
<h2><a name="bugs">Bugs</a></h2>
<ul>
<li><strong><a name="item_no_bugs_currently_known">No bugs currently known</a></strong><br />
</li>
If you find a bug with this software, file a bug report on the DAWG-PAWS
Sourceforge website: <a href="http://sourceforge.net/tracker/?group_id=204962">http://sourceforge.net/tracker/?group_id=204962</a>
<p></p></ul>
<p>
</p>
<h2><a name="limitations">Limitations</a></h2>
<ul>
<li><strong><a name="item_limited_support_for_triannot_output">Limited support for TriAnnot output</a></strong><br />
</li>
Conversion subfunctions are specific to the data type
being converted. Since the DAWG-PAWS process makes limited use of
the TriAnnot output, support is limited to the following
output files.
<ol>
<li><strong><a name="item_trf">trf</a></strong><br />
</li>
Tandem Repeat Finder
<p></p>
<li><strong><a name="item_eugos">eugOs</a></strong><br />
</li>
Eugene - Oryza sativa.
<p></p>
<li><strong><a name="item_fgh">fGh</a></strong><br />
</li>
Fgenesh - Gene Model Prediction
<p></p>
<li><strong><a name="item_gid">gID</a></strong><br />
</li>
GeneID - Gene Model Prediction.
<p></p>
<li><strong><a name="item_gmhv">gmHv</a></strong><br />
</li>
GeneMark Hordeum vulgare
<p></p>
<li><strong><a name="item_gmos">gmOs</a></strong><br />
</li>
GeneMark Oryza sativa.
<p></p>
<li><strong><a name="item_gmta">gmTa</a></strong><br />
</li>
GeneMark Triticum aestevum.
<p></p>
<li><strong><a name="item_gmzm">gmZm</a></strong><br />
</li>
GeneMark Zea mays
<p></p></ol>
</ul>
<p>
</p>
<hr />
<h1><a name="see_also">SEE ALSO</a></h1>
<p>The batch_blast.pl program is part of the DAWG-PAWS package of genome
annotation programs. See the DAWG-PAWS web page 
( <a href="http://dawgpaws.sourceforge.net/">http://dawgpaws.sourceforge.net/</a> )
or the Sourceforge project page 
( <a href="http://sourceforge.net/projects/dawgpaws">http://sourceforge.net/projects/dawgpaws</a> ) 
for additional information about this package.</p>
<p>
</p>
<hr />
<h1><a name="license">LICENSE</a></h1>
<p>GNU GENERAL PUBLIC LICENSE, VERSION 3</p>
<p><a href="http://www.gnu.org/licenses/gpl.html">http://www.gnu.org/licenses/gpl.html</a></p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<p>James C. Estill &lt;JamesEstill at gmail.com&gt;</p>
<p>
</p>
<hr />
<h1><a name="history">HISTORY</a></h1>
<p>STARTED: 02/09/2007

</p>
<p>UPDATED: 12/11/2007

</p>
<p>VERSION: $Rev: 330 $

</p>
<table border="0" width="100%" cellspacing="0" cellpadding="3">
<tr><td class="block" valign="middle">
<big><strong><span class="block">&nbsp;cnv_ta2ap.pl</span></strong></big>
</td></tr>
</table>

</body>

</html>
