<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
  <meta http-equiv="Content-Language" content="en-us">
  <meta http-equiv="Content-Type"
 content="text/html; charset=windows-1252">
  <title>DAWG-PAWS</title>
  <link rel="stylesheet" type="text/css" href="css/dawgpaws.css">
  <link rel="SHORTCUT ICON" href="images/dawgpaws.ico">
</head>
<body>
<table style="border-collapse: collapse;" id="AutoNumber1" border="0"
 bordercolor="#111111" cellpadding="0" cellspacing="0" width="100%">
  <tbody>
    <tr>
      <td width="50%">
      <p align="center"> <img src="images/DAWGPAWS_med.png"
 align="left" border="2" height="115" width="326"></p>
      </td>
      <td width="50%">
      <ul>
        <li><i><a href="#description">Description</a></i></li>
        <li><i><a href="#features">Features</a></i></li>
        <li><i><a href="#download">Download</a> <br>
          </i></li>
        <li><i><a href="http://www.gnu.org/licenses/gpl.html">License :
General Public License <br>
          </a></i></li>
        <li><i><a href="man.html">Manual Pages<br>
          </a></i></li>
        <li><i><a href="http://sourceforge.net/projects/dawgpaws/">SourceForge
Project Website</a></i></li>
        <li>Author: <a href="http://jestill.myweb.uga.edu">James
C. Estill</a></li>
      </ul>
      </td>
    </tr>
  </tbody>
</table>
<hr>
<h2><a name="description"></a>Description</h2>
<p>DAWG-PAWS is a suite of scripts written in Perl that are desinged to
assist a Distributed Annotation Working Group (<span
 style="font-weight: bold;">DAWG)</span> in the sequence annotation of
BAC
sized contigs. Since this suite of software was initially
written to annotate randomly sampled wheat BACs, it is refered to as a
Pipeline
to Annoate Wheat Sequnces (<span style="font-weight: bold;">PAWS</span>).
Although these program were initially designed for wheat, the scripts
can be applied to nearly any eukaryotic sequence annoation pipeline.<br>
</p>
<p>The scripts in this suite are designed to facilitate high throughput
generation of computational results for both gene annotation and
transposable element
annotation using a number of existing sequence
anotation programs. These results are converted from the native output
of the individual annotation programs to the standard <a
 href="http://www.sanger.ac.uk/Software/formats/GFF/">GFF&nbsp; file
format</a>. Since a GFF file is a simple tab delmited text file,
multiple computational results for a single sequence can easily be
concatenated together using standard text maninpulation tools such as <a
 href="http://linux.die.net/man/1/cat">cat</a>; proof that cats and
dawgs can play well together. DAWG-PAWS provides
additional tools to facilitate using the <a
 href="http://www.fruitfly.org/annot/apollo/">Apollo Genome Annotation
Curation Tool</a> for visualization and curation of the computational
results.<br>
</p>
<p>Usage of DAWG-PAWS is through a series of command line programs. A
command line interface was chosen since these programs will usually be
run on remote high preformance machines running some version of the
Unix/Linux operating system. Since a command line inteface has been
written
into all scripts, no special knowledge of Perl is required to use this
suite. However,
since some the Perl scripts rely on external Perl modules, a working
knowledge of Perl will be helpful when installing DAWG-PAWS.<br>
</p>
<h2><a name="features"></a>Features</h2>
<p><span lang="en-us">The current features of DAWG-PAWS include:</span></p>
<ul>
  <li>FASTA file manipulation to prepare sequences for the annotation
pipeline<br>
  </li>
  <li>Annotation of gaps in the sequence assemblies</li>
  <li>Executing ab initio gene anotation and transposable element
annotation
programs in a high throughput pipeline</li>
  <li>BLAST pipeline suitable for use in a cluster computing framework<br>
  </li>
  <li>Conversion of output from annotation software to the common GFF
file format</li>
  <li>Conversion of gff files to the game.xml format<br>
  </li>
  <li><span lang="en-us">Viewing game.xml results generated by
DAWG-PAWS in the Apollo genome annotation curation program<br>
    </span></li>
  <ul>
    <li><a
 href="http://dawgpaws.svn.sourceforge.net/viewvc/dawgpaws/trunk/apollo/wheat.tiers?view=log"><span
 lang="en-us">wheat.tiers</span></a></li>
    <li><span lang="en-us"><a
 href="http://dawgpaws.svn.sourceforge.net/viewvc/dawgpaws/trunk/apollo/wheat.style?view=log">wheat.style</a><br>
      </span></li>
  </ul>
  <li><span lang="en-us"><a
 href="http://tldp.org/HOWTO/Emacs-Beginner-HOWTO-3.html">Emacs modes</a>
for working with the apollo in the <a
 href="http://www.gnu.org/software/emacs/">Emacs text editing program</a>.<br>
    </span></li>
  <ul>
    <li><a
 href="http://dawgpaws.svn.sourceforge.net/viewvc/dawgpaws/trunk/emacs_modes/apollo-config-mode.el?view=log"><span
 lang="en-us">apollo-config-mode.el</span></a></li>
    <li><a
 href="http://dawgpaws.svn.sourceforge.net/viewvc/dawgpaws/trunk/emacs_modes/apollo-styles-mode.el?view=log"><span
 lang="en-us">apollo-styles-mode.el</span></a></li>
    <li><span lang="en-us"><a
 href="http://dawgpaws.svn.sourceforge.net/viewvc/dawgpaws/trunk/emacs_modes/apollo-tiers-mode.el?view=log">apollo-tiers-mode.el</a></span></li>
  </ul>
  <li><span lang="en-us">Help is available in all command line programs
using command line flags. Help info and full manual files are still
under
development for some programs. To access help where </span><big><code>progname</code></big><span
 lang="en-us"> is the name of the
individual program:<br>
    </span></li>
  <ul>
    <li><big><code>progname --usage </code></big><br>
generate a basic usage message with info on the required arguments<br>
    </li>
    <li><span lang="en-us"><big><code>progname --help</code></big><br>
generate a more extensive help message that includes required arguments
and all options<br>
      </span></li>
    <li><span lang="en-us"><big><code>progname --man</code></big><br>
view the full manual for the program</span></li>
  </ul>
  <li><span lang="en-us"><a href="man.html">Manual pages</a> for the
individual programs are
being made available on DAWG-PAWS web site as they are processed<br>
    </span></li>
</ul>
<h2><a name="download"></a>Download</h2>
<p>The full DAWG-PAWS suite will be available in a single download in
the future. Until then, the individual scripts may be downloaded
separately through the <a
 href="http://dawgpaws.svn.sourceforge.net/viewvc/dawgpaws/trunk/scripts/">SVN
browse interface</a>. The currently recommended way to get DAWG-PAWS is
to <a href="https://sourceforge.net/svn/?group_id=204962">anonymously
check out</a> the root directory from the SVN server. By downloading
this software you agree to abide by the
<a href="http://www.gnu.org/licenses/gpl.html">GNU General Public
License</a>.</p>
<span style="text-decoration: underline;"></span>
<ul>
</ul>
<h2><a name="manual"></a>User Manual and Software Documentation<span
 style="font-weight: 400;"></span></h2>
<p>A user manual describing the use of DAWG-PAWS for genome
annotation is currently under developement. This will be available for
download here when complete, until then you can access the manual that
is under development from the SVN. The manual pages for&nbsp;
individual
programs will be available as well. <br>
</p>
<h2>Supported Annotation Programs<br>
</h2>
<p>DAWG-PAWS includes scripts to work with computational evidences from
the following standalone programs and web sites. You can get
information for the following programs by clicking on the annotation
program name, and you can get the documentation for the DAWG-PAWS
script by clicking on the script name.<br>
</p>
<table style="text-align: left; width: 100%;" border="0" cellpadding="2"
 cellspacing="2">
  <tbody>
    <tr>
      <td colspan="2" rowspan="1"
 style="vertical-align: top; background-color: rgb(192, 192, 192);"><span
 style="font-weight: bold;">Gene Annotation</span></td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><span style="font-weight: bold;">Annotation
Program</span><br>
      </td>
      <td style="vertical-align: top;"><span style="font-weight: bold;">DAWG-PAWS
Scripts</span><br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><a
 href="http://www.ncbi.nlm.nih.gov/blast/Blast.cgi">NCBI-BLAST</a> </td>
      <td style="vertical-align: top;"><a href="man/batch_blast.html">batch_blast.pl</a><br>
      <a href="man/batch_cnv_blast2gff.html">batch_cnv_blast2gff.pl</a><br>
cnv_blast2gff.pl<br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><a
 href="http://urgi.versailles.inra.fr/projects/TriAnnot/">TriAnnot Web
Service</a><br>
      </td>
      <td style="vertical-align: top;">batch_cnv_ta2ap.pl<br>
cnv_ta2ap.pl<br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><a
 href="http://genes.mit.edu/GENSCAN.html">Genscan</a><br>
      </td>
      <td style="vertical-align: top;">run_genscan.pl<br>
batch_genscan.pl<br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><a
 href="http://www.softberry.com/berry.phtml">Fgenesh</a><br>
      </td>
      <td style="vertical-align: top;">run_fgenesh.pl<br>
cnv_fgenesh2gff.pl<br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><a
 href="http://exon.gatech.edu/GeneMark/">Genemark</a> </td>
      <td style="vertical-align: top;">batch_genemark.pl<br>
cnv_genemark2gff.pl </td>
    </tr>
    <tr>
      <td colspan="2" rowspan="1"
 style="vertical-align: top; background-color: rgb(192, 192, 192);"><span
 style="font-weight: bold;">Transposable Element Annotation</span><br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><span style="font-weight: bold;">Annotation
Program</span><br>
      </td>
      <td style="vertical-align: top;"><span style="font-weight: bold;">DAWG-PAWS
Scripts</span><br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><a
 href="http://www.repeatmasker.org/">RepeatMasker</a><br>
      </td>
      <td style="vertical-align: top;">batch_mask.pl<br>
batch_hardmask.pl<br>
cnv_repmask2gff.pl<br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><a
 href="http://www.ncbi.nlm.nih.gov/blast/Blast.cgi">NCBI-BLAST</a><br>
      </td>
      <td style="vertical-align: top;">batch_blast.pl<br>
batch_cnv_blast2gff.pl<br>
cnv_blast2gff.pl<br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><a
 href="http://www.public.iastate.edu/%7Eimagefpc/Subpages/software.html">TE
Nest</a><br>
      </td>
      <td style="vertical-align: top;">fetch_tenest.pl<br>
cnv_tenest2gff.pl<br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><a
 href="http://jaketu.biochem.vt.edu/4download/MITE/">FindMite</a><br>
      </td>
      <td style="vertical-align: top;">batch_findmite.pl<br>
cnv_findmite2gff.pl<br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><a
 href="http://darwin.informatics.indiana.edu/cgi-bin/evolution/ltr.pl">find_ltr</a><br>
      </td>
      <td style="vertical-align: top;">batch_findltr.pl<br>
cnv_findltr2gff.pl<br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><a
 href="http://www.eecs.wsu.edu/%7Eananth/software.htm">ltr_par (now
ltr_seq)</a><br>
      </td>
      <td style="vertical-align: top;">batch_ltrseq.pl<br>
cnv_ltrseq2gff.pl<br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><a
 href="http://www.genetics.uga.edu/retrolab/data/LTR_Struc.html">LTR_STRUC</a><br>
      </td>
      <td style="vertical-align: top;">ltrstruc_prep.pl<br>
batch_ltrstruc.vbs<br>
cnv_ltrstruc2gff.pl<br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><a
 href="http://hmmer.janelia.org/">HMMER</a><br>
      </td>
      <td style="vertical-align: top;">batch_hmmer.pl<br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><a
 href="http://tlife.fudan.edu.cn/ltr_finder">LTR_Finder</a> </td>
      <td style="vertical-align: top;">cnv_ltrfinder2gff.pl<br>
      </td>
    </tr>
  </tbody>
</table>
<hr>
<table style="border-collapse: collapse;" id="AutoNumber2" border="0"
 bordercolor="#111111" cellpadding="0" cellspacing="0" width="100%">
  <tbody>
    <tr>
      <td width="50%"><i>Author: <a
 href="http://www.arches.uga.edu/%7Ejestill">James Estill</a><br>
Last Updated: Dec 5, 2007</i></td>
      <td width="50%"> <br>
      </td>
      <td><a href="http://sourceforge.net/"> <img
 src="http://sourceforge.net/sflogo.php?group_id=204962&amp;type=5"
 alt="SourceForge.net Logo" border="0" height="62" width="210"></a> </td>
    </tr>
  </tbody>
</table>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<script src="http://www.google-analytics.com/urchin.js"
 type="text/javascript">
</script>
<script type="text/javascript">
_uacct = "UA-2207628-5";
urchinTracker();
</script>
</body>
</html>
